### PR TITLE
Refactor detailed diff v2 to short circuit on nulls and unknowns

### PR DESCRIPTION
**Pull request by VenelinMartinov**
This PR changes the logic in the detailed diff v2 calculation to short-circuit on encountering nils and unknowns.

Previously, when we encountered a null or unknown in either news or olds we would still recurse down and compare the non-nil values versus a nil value in order to get any replacements which might be happening. We would then simplify the diff later to trim these extra diffs but would propagate the replacement to the right level.

We would now instead short-circuit on encountering a null or unknown and walk the non-null value to find any properties which can trigger a replacement. This makes it much easier to handle sets in https://github.com/pulumi/pulumi-terraform-bridge/pull/2451 as recursing into sets is not necessary, as they are only compared by hashes.

This change is not meant to change any behaviour and is tested in https://github.com/pulumi/pulumi-terraform-bridge/pull/2405

Stacked on https://github.com/pulumi/pulumi-terraform-bridge/pull/2515 and https://github.com/pulumi/pulumi-terraform-bridge/pull/2516

### CODE REVIEW

File: pkg/tfbridge/detailed_diff.go [hunk 1]
```diff
@@ -41,6 +41,105 @@ func sortedMergedKeys[K cmp.Ordered, V any, M ~map[K]V](a, b M) []K {
...
+}
+
+// walkPropertyValue walks a property value and calls the visitor function for each path in the property value.
+func walkPropertyValue(
+	val resource.PropertyValue, path propertyPath, visitor func(propertyPath, resource.PropertyValue) bool,
+) bool {
+	if !visitor(path, val) {
+		return false
+	}
+
```
Comment (iwahbe) [inline pkg/tfbridge/detailed_diff.go:74]: Again, I'd like to see an assert that we are not hitting marker types.

File: pkg/tfbridge/detailed_diff.go [hunk 2]
```diff
@@ -41,6 +41,105 @@ func sortedMergedKeys[K cmp.Ordered, V any, M ~map[K]V](a, b M) []K {
...
+		for k, elVal := range val.ObjectValue() {
+			if !walkPropertyValue(elVal, path.Subpath(string(k)), visitor) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func willTriggerReplacement(
```
Comment (iwahbe) [inline pkg/tfbridge/detailed_diff.go:93]: Can you comment on why we need `willTriggerReplacement` and `willTriggerReplacementRecursive`.

File: pkg/tfbridge/detailed_diff.go [hunk 3]
```diff
@@ -41,6 +41,105 @@ func sortedMergedKeys[K cmp.Ordered, V any, M ~map[K]V](a, b M) []K {
...
+		return false
+	}
+	// Note this is mimicking the TF behaviour, so the effective type is not considered here.
+	if tfs.Type() != shim.TypeList && tfs.Type() != shim.TypeSet && tfs.Type() != shim.TypeMap {
+		return false
+	}
+	return isForceNew(tfs, ps)
+}
+
+func willTriggerReplacementRecursive(
```
Comment (iwahbe) [inline pkg/tfbridge/detailed_diff.go:126]: The name for this function needs to describe what it does, not how its' implemented.

File: pkg/tfbridge/detailed_diff.go [hunk 4]
```diff
@@ -41,6 +41,105 @@ func sortedMergedKeys[K cmp.Ordered, V any, M ~map[K]V](a, b M) []K {
...
+
+	parent := path[:len(path)-1]
+	tfs, ps, err = lookupSchemas(parent, rootTFSchema, rootPulumiSchema)
+	if err != nil {
+		return false
+	}
+	// Note this is mimicking the TF behaviour, so the effective type is not considered here.
+	if tfs.Type() != shim.TypeList && tfs.Type() != shim.TypeSet && tfs.Type() != shim.TypeMap {
+		return false
+	}
```
Comment (iwahbe) [inline pkg/tfbridge/detailed_diff.go:122]: I'm surprised that this is (1) only effected by one level up and (2) doesn't include ForceNew on objects.

File: pkg/tfbridge/detailed_diff.go [hunk 5]
```diff
@@ -318,27 +377,15 @@ func (differ detailedDiffer) makeListDiff(
...
-	}
-
 	return diff
 }
 
 func (differ detailedDiffer) makeMapDiff(
 	path propertyPath, old, new resource.PropertyValue,
 ) map[detailedDiffKey]*pulumirpc.PropertyDiff {
+	oldMap := old.ObjectValue()
+	newMap := new.ObjectValue()
```
Comment (iwahbe) [inline pkg/tfbridge/detailed_diff.go:387]: Just so I understand, this is ok because of `isTypeShapeMismatched`?

**Comment by iwahbe** [comment]
Reviewing the tests added in https://github.com/pulumi/pulumi-terraform-bridge/pull/2405, they look sufficient for nested forcenew in top level and singly nested cases. Can you point me to where multiple nesting levels are tested? I want to see a test that validates our behavior on a schema like this:

```
List{
  ForceNew: true,
  Elem: Object{
    Fields: {
       "f0": String,
    }
  },
}
```

I couldn't find tests for more then one level of nesting.

File: pkg/tfbridge/detailed_diff.go (hunk 4 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go:122]: I'm not sure I understand these - can you please expand on this?

Comment (iwahbe) [reply pkg/tfbridge/detailed_diff.go:122]: I'm surprised that we don't search all the way up the schema tree, not just one level up.

I think I was wrong about not including objects. `tfs.Type()` doesn't express objects.

File: pkg/tfbridge/detailed_diff.go (hunk 1 above)
Comment (t0yv0) [reply pkg/tfbridge/detailed_diff.go:74]: Just was commenting on another PR, but this is not a complete walker on PropertyValue.  

E.G.

https://github.com/pulumi/pulumi-terraform-bridge/blob/master/unstable/propertyvalue/propertyvalue.go#L42

If we had a complete walker on property value we would not need to double-test it. I know Go is about "a bit more duplication is better than a bit more sharing", but if we can reuse that makes coverage goals easier to achieve.

File: pkg/tfbridge/detailed_diff.go [hunk 6]
```diff
@@ -41,6 +41,105 @@ func sortedMergedKeys[K cmp.Ordered, V any, M ~map[K]V](a, b M) []K {
...
+			}
+		}
+	}
+	return true
+}
+
+func willTriggerReplacement(
+	path propertyPath, rootTFSchema shim.SchemaMap, rootPulumiSchema map[string]*info.Schema,
+) bool {
+	// A change on a property might trigger a replacement if:
```
Comment (t0yv0) [inline pkg/tfbridge/detailed_diff.go:96]: The function seems to do what the comment says. The rule sounds a little un-intuitive but looks like you've explored this thoroughly in tests so perhaps that's that? 

If a property P is forcenew, but P.a.b.c.d has changed, I am guessing that the overall plan is a replace plan, is that right? If we are displaying the detailed diff as `a.b.c.d has changed` we are probably within rights to blame `d` for replacing?

File: pkg/tfbridge/detailed_diff.go (hunk 4 above)
Comment (t0yv0) [reply pkg/tfbridge/detailed_diff.go:122]: Yeah I was wondering this too. We can probably get a away with something approximate as it's presentation-layer only but worth backlogging at least.

File: pkg/tfbridge/detailed_diff.go [hunk 7]
```diff
@@ -264,13 +294,48 @@ func (differ detailedDiffer) makePlainPropDiff(
 	return nil
 }
 
+// makeShortCircuitDiff is used for properties that are nil or computed in either the old or new state.
```
Comment (t0yv0) [inline pkg/tfbridge/detailed_diff.go:297]: This is .. tricky.. If a property is Computed or unknown, we do not know if it will trigger a replacement or not, until we know it. It might trigger a replacement, or it might not. Any implementation here takes a guess and we need to probably call this out in a comment that we are guessing. I would look up current Pulumi behavior for this case and copy how we're guessing to continue in the tradition we have.

File: pkg/tfbridge/detailed_diff_test.go [hunk 8]
```diff
@@ -42,11 +42,9 @@ func TestSchemaLookupMaxItemsOnePlain(t *testing.T) {
 		},
 	}
 
-	differ := detailedDiffer{
-		tfs: shimv2.NewSchemaMap(sdkv2Schema),
-	}
+	tfs := shimv2.NewSchemaMap(sdkv2Schema)
```
Comment (t0yv0) [inline pkg/tfbridge/detailed_diff_test.go:45]: The changes to tests look very unconvincing 🙏 

You're going for no change in behavior in this PR? 

Can you point me to a test case that illustrates this "short-circuit" diff so I can build up some intuition? Perhaps this all is under tests already. Thank you!

Do not want to block on this but would like to understand better.

File: pkg/tfbridge/detailed_diff.go (hunk 7 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go:276]: Added tests in https://github.com/pulumi/pulumi-terraform-bridge/pull/2515. The existing behaviour looks wrong in some cases.

File: pkg/tfbridge/detailed_diff.go [hunk 9]
```diff
@@ -41,6 +41,105 @@ func sortedMergedKeys[K cmp.Ordered, V any, M ~map[K]V](a, b M) []K {
 	return keysSlice
 }
 
+func isTypeShapeMismatched(val resource.PropertyValue, propType shim.ValueType) bool {
+	if !isPresent(val) {
+		return false
+	}
+	switch propType {
```
[the remark this answers is not in the document]
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go:49]: I believe IsOutput implies IsComputed, so both should not be necessary.

File: pkg/tfbridge/detailed_diff.go (hunk 2 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go:68]: I've renamed these, hope it's clearer.

willTriggerReplacement only checks the property at the given path. willTriggerReplacementRecursive recurses deeply into a propertyValue to check if any of its properties will trigger the replacement.

File: pkg/tfbridge/detailed_diff.go (hunk 4 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go:97]: I believe the behaviour here is correct as I've checked it against terraform.

A property will trigger a replacement if:
1. It has ForceNew
2. It is an element of a collection which has ForceNew
3. It is somewhere in a set element which is changed and contains a ForceNew property - set elements are always replaced.

File: pkg/tfbridge/detailed_diff.go (hunk 5 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go:366]: yeah, I moved all the asserts there

**Comment by VenelinMartinov** [comment]
>Reviewing the tests added in https://github.com/pulumi/pulumi-terraform-bridge/pull/2405, they look sufficient for nested forcenew in top level and singly nested cases. Can you point me to where multiple nesting levels are tested? I want to see a test that validates our behavior on a schema like this:

Here's the tests for ForceNew:

https://github.com/pulumi/pulumi-terraform-bridge/blob/629a5ff3544665821a49127753511deece30fe8e/pkg/tfbridge/detailed_diff_test.go#L859
https://github.com/pulumi/pulumi-terraform-bridge/blob/629a5ff3544665821a49127753511deece30fe8e/pkg/tfbridge/detailed_diff_test.go#L923
https://github.com/pulumi/pulumi-terraform-bridge/blob/629a5ff3544665821a49127753511deece30fe8e/pkg/tfbridge/detailed_diff_test.go#L1057
https://github.com/pulumi/pulumi-terraform-bridge/blob/629a5ff3544665821a49127753511deece30fe8e/pkg/tfbridge/detailed_diff_test.go#L1165
https://github.com/pulumi/pulumi-terraform-bridge/blob/629a5ff3544665821a49127753511deece30fe8e/pkg/tfbridge/detailed_diff_test.go#L1401

I believe I've covered all shapes and nestings of TF schemas there.

I've also added additional tests which focus on how forceNew interacts with Unknowns in https://github.com/pulumi/pulumi-terraform-bridge/pull/2515

File: pkg/tfbridge/detailed_diff.go (hunk 6 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go:71]: That's not quite the case as any object along the way stops the propagation of the change.

in your example if b is an object and d has changed then TF will list b as updated instead and not trigger a replace on P.

See also https://github.com/pulumi/pulumi-terraform-bridge/pull/2496#discussion_r1819350356

Cross tests for this in https://github.com/pulumi/pulumi-terraform-bridge/blob/49aaf757dc71785b2dd60ee0af249f794f19235f/pkg/internal/tests/cross-tests/diff_cross_test.go#L1369

File: pkg/tfbridge/detailed_diff_test.go (hunk 8 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff_test.go:45]: The refactor moved the lookupSchemas function to be independent of the differ class, hence these changes.

Which part is unconvincing? The behaviour is meant to be under tests, hence why there are no additional changes here - the existing tests still work fine.


The short-circuiting behaviour is triggered when the two propertyValue trees we are diffing are not balanced - e.g. we see a nil versus a collection or a nil vs an object. Previously we'd recurse into the object and compare its properties against nil but now we short-circuit on encountering a nil. This enabled the whole thing to work for sets where we can not recurse into set elements.

File: pkg/tfbridge/detailed_diff.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (t0yv0) [reply pkg/tfbridge/detailed_diff.go:50]: FWIW Computed and Output are distinct cases.

```
// IsComputed returns true if the underlying value is a computed value.
func (v PropertyValue) IsComputed() bool {
	_, is := v.V.(Computed)
	return is
}

// IsOutput returns true if the underlying value is an output value.
func (v PropertyValue) IsOutput() bool {
	_, is := v.V.(Output)
	return is
}

```

File: pkg/tfbridge/detailed_diff.go [hunk 10]
```diff
@@ -41,6 +42,84 @@ func sortedMergedKeys[K cmp.Ordered, V any, M ~map[K]V](a, b M) []K {
...
+	}
+	return isForceNew(tfs, ps)
+}
+
+func propertyValueTriggersReplacement(
+	path propertyPath, value resource.PropertyValue, tfs shim.SchemaMap, ps map[string]*info.Schema,
+) bool {
+	replacement := false
+	visitor := func(subpath resource.PropertyPath, val resource.PropertyValue) (resource.PropertyValue, error) {
+		if propertyPathTriggersReplacement(propertyPath(subpath), tfs, ps) {
```
Comment (t0yv0) [inline pkg/tfbridge/detailed_diff.go:107]: IS this right? Should it be `path+subpath`?

File: pkg/tfbridge/detailed_diff.go [hunk 11]
```diff
@@ -264,13 +274,48 @@ func (differ detailedDiffer) makePlainPropDiff(
...
+func (differ detailedDiffer) makeShortCircuitDiff(
+	path propertyPath, old, new resource.PropertyValue,
+) map[detailedDiffKey]*pulumirpc.PropertyDiff {
+	contract.Assertf(old.IsNull() || new.IsNull() || new.IsComputed(),
+		"short-circuit diff should only be used for nil properties")
+	if old.IsNull() && new.IsNull() {
+		return nil
+	}
+
+	baseDiff := makeBaseDiff(old, new)
```
Comment (t0yv0) [inline pkg/tfbridge/detailed_diff.go:288]: Should this handle NoDiff?

Comment (t0yv0) [reply pkg/tfbridge/detailed_diff.go:288]: Ah I guess it's the degenerate case. There must be a diff here.

File: pkg/tfbridge/detailed_diff.go [hunk 12]
```diff
@@ -264,13 +274,48 @@ func (differ detailedDiffer) makePlainPropDiff(
...
+	contract.Assertf(baseDiff != undecidedDiff, "short-circuit diff could not determine diff kind")
+
+	propDiff := baseDiff.ToPropertyDiff()
+	if new.IsComputed() && propertyPathTriggersReplacement(path, differ.tfs, differ.ps) {
+		propDiff = promoteToReplace(propDiff)
+	} else if !new.IsNull() && !new.IsComputed() && propertyValueTriggersReplacement(path, new, differ.tfs, differ.ps) {
```
Comment (t0yv0) [inline pkg/tfbridge/detailed_diff.go:294]: Had to think about this one. This is suspect. I'm not sure this is right, not sure how to verify this. This analyzes schemas for all sub-paths realized in `new` property Value, without any consideration as to whether these sub-paths result in diff or no-diff compared to the `old` value. Why is that? Ah that's because this assumes old.IsNull.

File: pkg/tfbridge/detailed_diff_test.go (hunk 8 above)
Comment (t0yv0) [reply pkg/tfbridge/detailed_diff_test.go:45]: The change is large and I'm pretty sure changes behavior. I don't think this is a refactor given the test cov we have.

File: pkg/tfbridge/detailed_diff.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (t0yv0) [reply pkg/tfbridge/detailed_diff.go:50]: ``` contract.Assertf(!val.IsComputed() && !val.IsSecret(), "val should not be computed or secret") ```

Asserts are good to have but I don't know why we're not hitting them. Sounds like first-class outputs or secrets for example are bound to show up sometime in the code this is diffing? Why do we think the asserts will not hit? 

What about first-class secrets? If these do not show up it's for some very complicated historical reasons around secrets handling. 

Would it be cheaper just to handle these forms and get to 100% rather than keep chasing this down later on? 

The case analysis is as such. 

1. A value in news may be an unknown. This may be represented as Computed or Output that is unknown

2. A value in either olds or news may be a Secret or an Output with secret=true. For the purposes of this we can drop the secret bit-ness and continue computing differing paths as before, I think. Might worth thinking through this one.

3. Finally a value may be an Output that is secret=false, known=true. Unwrap and proceed on the inner value.

File: pkg/tfbridge/detailed_diff.go [hunk 13]
```diff
@@ -264,13 +274,48 @@ func (differ detailedDiffer) makePlainPropDiff(
...
 	path propertyPath, old, new resource.PropertyValue,
 ) map[detailedDiffKey]*pulumirpc.PropertyDiff {
 	if path.IsReservedKey() {
 		return nil
 	}
 	propType := differ.getEffectiveType(differ.propertyPathToSchemaPath(path))
+	if !isPresent(old) || isTypeShapeMismatched(old, propType) {
+		old = resource.NewNullProperty()
+	}
+	if !new.IsComputed() && (!isPresent(new) || isTypeShapeMismatched(new, propType)) {
```
[the remark this answers is not in the document]
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go:313]: I can handle this more carefully but I've assumed this to be a degenerate case and have handled it as such.

File: pkg/tfbridge/detailed_diff.go (hunk 12 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go:294]: Do you think I should add a comment here? Which part is suspect?

File: pkg/tfbridge/detailed_diff.go (hunk 10 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go:107]: Pretty sure subpath contains path, will double check here.

File: pkg/tfbridge/detailed_diff_test.go (hunk 8 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff_test.go:45]: The change is big but I have added quite a few tests here, pretty sure the coverage is good. The follow-up work also works quite well, so I am fairly confident in this change.

What do you think is missing?

**Comment by t0yv0** [comment]
Here's a case where it just assumes nil (no changes) because "the shape mismatched". Why?

```go
func TestAddingSecretArray(t *testing.T) {
	old := resource.PropertyMap{"x": resource.NewNullProperty()}
	new := resource.PropertyMap{"x": resource.MakeSecret(resource.NewArrayProperty([]resource.PropertyValue{
		resource.NewStringProperty("SECRET"),
	}))}
	tfs := shimv2.NewSchemaMap(map[string]*schema.Schema{
		"x": &schema.Schema{
			Type: schema.TypeList,
			Elem: &schema.Schema{
				Type: schema.TypeString,
			},
			Optional: true,
		},
	})
	runDetailedDiffTest(t, old, new, tfs, nil, map[string]*pulumirpc.PropertyDiff{
		"x": &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD},
	})
}

```

**Comment by VenelinMartinov** [comment]
@t0yv0 the test you posted seems to work fine. It returns an {x: ADD} which is the expected value.

I guess we might get it off in cases like these:
```go
func TestMistypedArray(t *testing.T) {
	old := resource.PropertyMap{"x": resource.NewStringProperty("a")}
	new := resource.PropertyMap{"x": resource.NewArrayProperty([]resource.PropertyValue{
		resource.NewStringProperty("VAL"),
	})}
	tfs := shimv2.NewSchemaMap(map[string]*schema.Schema{
		"x": &schema.Schema{
			Type: schema.TypeList,
			Elem: &schema.Schema{
				Type: schema.TypeString,
			},
			Optional: true,
		},
	})
	runDetailedDiffTest(t, old, new, tfs, nil, map[string]*pulumirpc.PropertyDiff{
		"x[0]": &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD},
	})
}

```

Here we present an ADD diff when it is really an UPDATE. I think this is enough of a corner case and the diff is still comprehensible that it's probably fine.

**Comment by t0yv0** [comment]
Hmm, interesting, do I have a problem with local setup? The test fails for me.

**Comment by t0yv0** [comment]
> Additionally the behaviour was there before, I am fairly sure this PR does not change that.

So we have a bug, yes? 

If I replace code with this:

```
func isTypeShapeMismatched(val resource.PropertyValue, propType shim.ValueType) bool {
	return false
}
```

All the tests still pass.

**Comment by VenelinMartinov** [comment]
I'll add some tests to make the behaviour clearer

**Comment by t0yv0** [comment]
I've pulled your latest changes, and now TestAddingSecretArray panics.

**Comment by t0yv0** [comment]
Yes, but, we will need to change that  to support secrets properly, and at a minimum we need the work backlogged so we know there are additional bugs in diff code.
 

I think I'd like to see the code factored out and made a little simpler until we have confidence that we're not adding panics into the product or replacing diffs with nil diffs.

Your TestMistypedArray is another bug example. To reduce confidence further, I get a different result on my machine than claimed. IT produces an `"x": ADD` diff. It really an UPDATE as you point out.

What are the next steps here, are we fixing the code or logging known bugs?

**Comment by VenelinMartinov** [comment]
Handling secrets is out of scope here - what panics is an assert saying we don't support secrets in detailed diff.

I'll add an item to add additional tests for mistyped state so that we can decide if this is worth including in the epic. This is out of scope for this PR.

File: pkg/tfbridge/detailed_diff.go (hunk 12 above)
Comment (t0yv0) [reply pkg/tfbridge/detailed_diff.go:294]: I grokked it but it required effort.

File: pkg/tfbridge/detailed_diff.go [hunk 14]
```diff
@@ -307,10 +307,15 @@ func (differ detailedDiffer) makePropDiff(
 		return nil
 	}
 	propType := differ.getEffectiveType(differ.propertyPathToSchemaPath(path))
-	if !isPresent(old) || isTypeShapeMismatched(old, propType) {
+
+	if isTypeShapeMismatched(old, propType) || isTypeShapeMismatched(new, propType) {
```
Comment (t0yv0) [inline pkg/tfbridge/detailed_diff.go:311]: This makes a lot more sense to me, ty!

File: pkg/tfbridge/detailed_diff.go [hunk 15]
```diff
@@ -181,7 +181,7 @@ func makeBaseDiff(old, new resource.PropertyValue) baseDiff {
 		return deleteDiff
 	}
 
-	if new.IsComputed() {
+	if new.IsComputed() || new.IsOutput() {
```
Comment (t0yv0) [inline pkg/tfbridge/detailed_diff.go:187]: This line is not right. Output may be known. I think that you might get away with this (since Bridge claims it does not support first-class outputs).

```
contract.Assertf(!val.IsOutput())
```

File: pkg/tfbridge/detailed_diff.go [hunk 16]
```diff
@@ -279,8 +279,8 @@ func (differ detailedDiffer) makePlainPropDiff(
 func (differ detailedDiffer) makeShortCircuitDiff(
 	path propertyPath, old, new resource.PropertyValue,
 ) map[detailedDiffKey]*pulumirpc.PropertyDiff {
-	contract.Assertf(old.IsNull() || new.IsNull() || new.IsComputed(),
-		"short-circuit diff should only be used for nil properties")
+	contract.Assertf(old.IsNull() || new.IsNull() || new.IsComputed() || new.IsOutput(),
```
Comment (t0yv0) [inline pkg/tfbridge/detailed_diff.go:285]: This isn't right either.

**Comment by t0yv0** [comment]
I've added https://github.com/pulumi/pulumi-terraform-bridge/issues/2525 that'd give some basic confidence that these asserts do not result in panics.

**Review by t0yv0 — APPROVED** [review]
This is better. LGTM now, but don't try to handle Outputs incorrectly, rather panic, I think  we've looked and they do not show up in the bridge protocol I think. 

If there's any know remaining bugs in this code, please file, otherwise lgtm.

File: pkg/tfbridge/detailed_diff.go (hunk 15 above)
Comment (VenelinMartinov) [reply pkg/tfbridge/detailed_diff.go]: It is not right but this is how we handle outputs in schema.go - I've reverted the change however
